### ISSUE TITLE
Lazy Loading Failing to Load

### ISSUE THREAD

**Issue by SuricateCan** [aspnet/EntityFrameworkCore#17745]
Today I stumbled across a weird bug that prevented a navigation property to be loaded.

In my first scenario, the navigation property was one-to-many with no foreign key property specified, so, a shadow property was created for it.
Turning on the debug on EF classes, it simply says "Lazy-loading property" and just skips to the next one.
After spending some time debugging EF Core, I found out that the [`_shadowValues`](https://github.com/aspnet/EntityFrameworkCore/blob/24b9aa1d2e14fe2e737255ede9b2a7a623fcf2af/src/EFCore/ChangeTracking/Internal/InternalMixedEntityEntry.cs#L21) did not contain the foreign key value (which was surely set), resulting in the navigation property not loading.

Note that I said "first scenario".

After some time, I managed to isolate the issue and gather some important information about the behavior, which I describe below.

### Steps to reproduce

I've created a [gist](https://gist.github.com/SuricateCan/4063fe98ca1cee5a21a5bc10b14d4dcd) containing a cleaned up code that can reproduce the behavior.
I suggest downloading the whole thing, placing the migrations inside a `Migrations` folder and building to run the tests.

#### Scenario 1 

Using the project exactly as it is, when running [`Program.cs`](https://gist.github.com/SuricateCan/4063fe98ca1cee5a21a5bc10b14d4dcd#file-program-cs) the result shows:

```text
Concrete1.Foo Working
Concrete2.Reference Not Working
```

This means that the referenced property could not be loaded, but `Foo` did.

#### Scenario 2

By accident, I discovered that I could change the behavior by refactoring my class name.
To do so, do as follows:

- Remove the current migration
- Refactor `Concrete2` to `Concrete0`
- Create a new migration
- Run [`Program.cs`](https://gist.github.com/SuricateCan/4063fe98ca1cee5a21a5bc10b14d4dcd#file-program-cs) 

The result is as follows:

```text
Concrete1.Foo Not Working
Concrete0.Reference Working
```

Note that the entity names affected their registration, movig the issue to another class. But the issue persisted.

### Workaround

I also find out a simple (yet annoying) workaround for this behavior.
If I uncomment the following lines:
[Model.cs:22](https://gist.github.com/SuricateCan/4063fe98ca1cee5a21a5bc10b14d4dcd#file-model-cs-L22)
```csharp
        //public int ReferenceId { get; set; }
```
[Context.cs:35-36](https://gist.github.com/SuricateCan/4063fe98ca1cee5a21a5bc10b14d4dcd#file-context-cs-L35-L36)
```csharp
            //modelBuilder.Entity<Concrete2>()
            //    .HasOne(a => a.Reference).WithMany().HasForeignKey(a => a.ReferenceId);
```
and generate a new migration, the issue dissapears completely. The property is correctly loaded, meaning that if I don't use a shadow property for FK, EF Core behaves as usual.

### Further technical details

EF Core version: 2.2.6
Database provider: Tested with `Microsoft.EntityFrameworkCore.SqlServer` and `Microsoft.EntityFrameworkCore.Sqlite`
Target framework:  .NET Core 2.2
Operating system: Windows 10 Enterprise
IDE: Visual Studio 2019 16.2.4

**EDIT:** The Snapshot showed a single difference after uncommenting the lines:

[ContextModelSnapshot.cs:59](https://gist.github.com/SuricateCan/4063fe98ca1cee5a21a5bc10b14d4dcd#file-contextmodelsnapshot-cs-L59)
```diff
-                   b.Property<int?>("ReferenceId");
+                   b.Property<int>("ReferenceId");
```

**Comment by ajcvickers** [comment]
@SuricateCan Thanks for the great bug filing and repro code! I was able to reproduce the issue on 2.2, but updating the EF packages resolved the issue. I don't know exactly what changed to fix this in 3.0, but there was a significant change to some of the lazy-loading infrastructure so I'm guessing it was fixed as part of that.

**Comment by SuricateCan** [comment]
@ajcvickers Thanks for the feedback.
Unfortunately, I'm unable to migrate to 3.0 right now.
Is there any plans to fix it for 2.2?

**Comment by ajcvickers** [comment]
@SuricateCan Only high-priority issues will be patched in previous releases. This issue doesn't meet that bar.
